subxt/src/events/event_subscription.rs

```diff
@@ -51,36 +57,111 @@ pub use super::{
 /// [`Events::subscribe_finalized()`] if that is important.
 ///
 /// **Note:** This function is hidden from the documentation
-/// and is exposed only to be called via the codegen. Thus, prefer to use
-/// `api.events().subscribe()` over calling this directly.
+/// and is exposed only to be called via the codegen. It may
+/// break between minor releases.
 #[doc(hidden)]
-pub async fn subscribe<T: Config, Evs: Decode + 'static>(
-    client: &'_ Client<T>,
-) -> Result<EventSubscription<'_, T, Evs>, BasicError> {
+pub async fn subscribe<'a, T: Config, Evs: Decode + 'static>(
+    client: &'a Client<T>,
+) -> Result<EventSubscription<'a, EventSub<T::Header>, T, Evs>, BasicError> {
     let block_subscription = client.rpc().subscribe_blocks().await?;
     Ok(EventSubscription::new(client, block_subscription))
 }
 
 /// Subscribe to events from finalized blocks.
 ///
 /// **Note:** This function is hidden from the documentation
-/// and is exposed only to be called via the codegen. Thus, prefer to use
-/// `api.events().subscribe_finalized()` over calling this directly.
+/// and is exposed only to be called via the codegen. It may
+/// break between minor releases.
 #[doc(hidden)]
-pub async fn subscribe_finalized<T: Config, Evs: Decode + 'static>(
-    client: &'_ Client<T>,
-) -> Result<EventSubscription<'_, T, Evs>, BasicError> {
-    let block_subscription = client.rpc().subscribe_finalized_blocks().await?;
-    Ok(EventSubscription::new(client, block_subscription))
+pub async fn subscribe_finalized<'a, T: Config, Evs: Decode + 'static>(
+    client: &'a Client<T>,
+) -> Result<EventSubscription<'a, FinalizedEventSub<'a, T::Header>, T, Evs>, BasicError> {
+    // fetch the last finalised block details immediately, so that we'll get
+    // events for each block after this one.
+    let last_finalized_block_hash = client.rpc().finalized_head().await?;
+    let last_finalized_block_number = client
+        .rpc()
+        .header(Some(last_finalized_block_hash))
+        .await?
+        .map(|h| (*h.number()).into());
+
+    // Fill in any gaps between the block above and the finalized blocks reported.
+    let block_subscription = subscribe_to_block_headers_filling_in_gaps(
+        client,
+        last_finalized_block_number,
+        client.rpc().subscribe_finalized_blocks().await?,
+    );
+
+    Ok(EventSubscription::new(client, Box::pin(block_subscription)))
 }
 
+/// Take a subscription that returns block headers, and if any block numbers are missed out
+/// betweem the block number provided and what's returned from the subscription, we fill in
+/// the gaps and get hold of all intermediate block headers.
+///
+/// **Note:** This is exposed so that we can run integration tests on it, but otherwise
+/// should not be used directly and may break between minor releases.
+#[doc(hidden)]
+pub fn subscribe_to_block_headers_filling_in_gaps<'a, S, E, T: Config>(
+    client: &'a Client<T>,
+    mut last_block_num: Option<u64>,
```

Comment:
Is it ok to assume that the block number is `u64` here? Or would `Option<Into<u64>>` work?

Reply:
since block numbers can always be converted into `u64`s now, it's just easier to do the conversion first and then pass in the u64's here (this function isn't "public" anyway).